diff --git a/basic_system/src/system_functions/mod.rs b/basic_system/src/system_functions/mod.rs
index de75e4f8..b7eae8b6 100644
--- a/basic_system/src/system_functions/mod.rs
+++ b/basic_system/src/system_functions/mod.rs
@@ -65,33 +65,6 @@ impl<R: Resources, const USE_ADVICE: bool> SystemFunctionsExt<R>
 {
     type Secp256k1ECRecover = ecrecover::EcRecoverImpl<USE_ADVICE>;
     type ModExp = modexp::ModExpImpl<USE_ADVICE>;
-
-    fn u256_div_rem<O: zk_ee::oracle::IOOracle>(
-        dividend_or_quotient: &mut u256::U256,
-        divisor_or_remainder: &mut u256::U256,
-        oracle: &mut O,
-    ) {
-        if USE_ADVICE {
-            u256_advice::u256_div_rem_with_advice(
-                dividend_or_quotient,
-                divisor_or_remainder,
-                oracle,
-            )
-        } else {
-            u256::U256::div_rem(dividend_or_quotient, divisor_or_remainder)
-        }
-    }
-
-    fn u256_mulmod<O: zk_ee::oracle::IOOracle>(
-        a: &mut u256::U256,
-        b: &mut u256::U256,
-        modulus_or_result: &mut u256::U256,
-        oracle: &mut O,
-    ) {
-        if USE_ADVICE {
-            u256_advice::u256_mulmod_with_advice(a, b, modulus_or_result, oracle)
-        } else {
-            u256::U256::mul_mod(a, b, modulus_or_result)
-        }
-    }
+    type DivRem = u256_advice::DivRemImpl<USE_ADVICE>;
+    type Mulmod = u256_advice::MulmodImpl<USE_ADVICE>;
 }
diff --git a/basic_system/src/system_functions/u256_advice.rs b/basic_system/src/system_functions/u256_advice.rs
index d0782651..bd1d16ca 100644
--- a/basic_system/src/system_functions/u256_advice.rs
+++ b/basic_system/src/system_functions/u256_advice.rs
@@ -2,11 +2,44 @@ use u256::U256;
 use zk_ee::oracle::query_ids::{U256_DIV_REM_ADVICE_QUERY_ID, U256_MULMOD_ADVICE_QUERY_ID};
 use zk_ee::oracle::usize_serialization::UsizeDeserializable;
 use zk_ee::oracle::IOOracle;
+use zk_ee::system::base_system_functions::{DivRemExt, MulmodExt};
 #[cfg(target_pointer_width = "32")]
 use zk_ee::utils::u256_arithmetic_advice::{U256DivRemAdviceParams, U256MulmodAdviceParams};
 #[cfg(target_pointer_width = "64")]
 use zk_ee::utils::u256_arithmetic_advice::{U256DivRemAdviceParams64, U256MulmodAdviceParams64};
 
+pub struct DivRemImpl<const USE_ADVICE: bool>;
+pub struct MulmodImpl<const USE_ADVICE: bool>;
+
+impl<const USE_ADVICE: bool> DivRemExt for DivRemImpl<USE_ADVICE> {
+    fn execute<O: IOOracle>(
+        dividend_or_quotient: &mut U256,
+        divisor_or_remainder: &mut U256,
+        oracle: &mut O,
+    ) {
+        if USE_ADVICE {
+            u256_div_rem_with_advice(dividend_or_quotient, divisor_or_remainder, oracle)
+        } else {
+            U256::div_rem(dividend_or_quotient, divisor_or_remainder)
+        }
+    }
+}
+
+impl<const USE_ADVICE: bool> MulmodExt for MulmodImpl<USE_ADVICE> {
+    fn execute<O: IOOracle>(
+        a: &mut U256,
+        b: &mut U256,
+        modulus_or_result: &mut U256,
+        oracle: &mut O,
+    ) {
+        if USE_ADVICE {
+            u256_mulmod_with_advice(a, b, modulus_or_result, oracle)
+        } else {
+            U256::mul_mod(a, b, modulus_or_result)
+        }
+    }
+}
+
 #[inline(always)]
 fn read_limbs_from_oracle_response(it: &mut impl ExactSizeIterator<Item = usize>) -> [u64; 4] {
     [
diff --git a/zk_ee/src/system/base_system_functions.rs b/zk_ee/src/system/base_system_functions.rs
index f8cde5f5..51a57b8e 100644
--- a/zk_ee/src/system/base_system_functions.rs
+++ b/zk_ee/src/system/base_system_functions.rs
@@ -388,6 +388,8 @@ pub trait SystemFunctions<R: Resources> {
 pub trait SystemFunctionsExt<R: Resources> {
     type Secp256k1ECRecover: SystemFunctionExt<R, Secp256k1ECRecoverErrors>;
     type ModExp: SystemFunctionExt<R, ModExpErrors>;
+    type DivRem: DivRemExt;
+    type Mulmod: MulmodExt;
 
     fn secp256k1_ec_recover<
         O: IOOracle,
@@ -425,12 +427,33 @@ pub trait SystemFunctionsExt<R: Resources> {
         dividend_or_quotient: &mut u256::U256,
         divisor_or_remainder: &mut u256::U256,
         oracle: &mut O,
-    );
+    ) {
+        Self::DivRem::execute(dividend_or_quotient, divisor_or_remainder, oracle)
+    }
 
     fn u256_mulmod<O: IOOracle>(
         a: &mut u256::U256,
         b: &mut u256::U256,
         modulus_or_result: &mut u256::U256,
         oracle: &mut O,
+    ) {
+        Self::Mulmod::execute(a, b, modulus_or_result, oracle)
+    }
+}
+
+pub trait DivRemExt {
+    fn execute<O: IOOracle>(
+        dividend_or_quotient: &mut u256::U256,
+        divisor_or_remainder: &mut u256::U256,
+        oracle: &mut O,
+    );
+}
+
+pub trait MulmodExt {
+    fn execute<O: IOOracle>(
+        a: &mut u256::U256,
+        b: &mut u256::U256,
+        modulus_or_result: &mut u256::U256,
+        oracle: &mut O,
     );
 }
